### ISSUE TITLE
support for Estuary

### ISSUE THREAD

**Issue by jegue** [rclone/rclone#5854]
#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by jegue** [comment]
https://docs.estuary.tech/

**Comment by Masamune3210** [comment]
Could you please go back and actually fill out the template

**Comment by marshall** [comment]
hello, chiming in here from the FIlecoin Foundation. for some added context, Estuary is a cloud storage backend that automates data migration into both IPFS and Filecoin, making it easier to use these decentralized networks.

I've been working on integrating rclone with Estuary in a fork maintained by the Filecoin application research group here:
https://github.com/application-research/rclone (see the`estuary` branch).

If there is enough interest from the rclone team, we would be happy to go through whatever formal process is needed to get it in shape for landing in the official repo. We definitely would prefer not to maintain a fork!

**Comment by ncw** [comment]
@marshall great work!

I had a quick skim over the code - looks good!

Check out the docs here for what needs to happen for a backend

https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend . The first step is to get the integration tests to pass.

We'd also need an account, or a docker image to test the backend against for the [daily integration tests](https://pub.rclone.org/integration-tests/current/)

We'd also want to have you help maintain the backend going forwards. We'd be happy to deal with the day to day stuff but we'd need an expert on tap for those hard questions!

**Comment by gmelodie** [comment]
@ncw hey there! So I'll be (trying to) taking marshall's place in owning the estuary fork and I've got a couple questions for us to move forward

**Comment by gmelodie** [comment]
@ncw ping!

**Comment by ncw** [comment]
> @ncw hey there! So I'll be (trying to) take marshall's place in owning the estuary fork and I've got a question: how do I disable tests that don't make sense? For instance the `mkdir` test shouldn't be run since it's not possible to create empty directories in estuary.

You should try to make as many tests pass as possible. s3 doesn't support empty directories either so mkdir is essentially a no-op...

> About having a docker image. Although that would be possible I think having to update it would be much harder than just providing you with an API key so we could test against the production server, I can get you one of those. Can you let me know how that is submited to you?

Email to nick@craig-wood.com is fine :-)

> Obs: I'm super happy with maintaining moving forward, as long as you never call me an expert... on anything haha stuck_out_tongue

Life is a learning process!

It you can get all the tests to pass then it will be a rock solid backend.

**Comment by ncw** [comment]
> 1. `TestIntegration/FsMkdir/FsMkdirRmdirSubdir` fails because it tries to create subdirs before removing them, and our backend doesn't allow for subdir creation, so I'm getting errors like these when running the tests:

If your backend treats directories the same way S3 does, namely they only exist when they have files in, then set the `BucketBased: true` feature flag.



> 2\. All `TestIntegration/FsMkdir/FsEncoding/` tests are failing apparently because of modification time being `0001-01-01 00:00:00 +0000 UTC`. I have no idea why that could be.

Do the other times work properly? If it is just the encoding ones then I'd suspect a bug in either the backend or Estuary.

**Comment by anjor** [comment]
Hey @ncw sorry about the radio silence from our end.

I fixed the modTime issues, needed some wrangling on our end. Quite a few tests are passing at this point. The one that is failing and I can't do much to fix it is `FsPutFiles` -- the test puts the same file name in the same path twice, expecting the second call to be an update; however the estuary backend does not support this behaviour. It throws an error when a user tries to upload duplicate files to the same location. 

There are a couple of tests in `FsEncoding` that I am still chasing down but generally things are looking much better.

**Comment by ncw** [comment]
Good news! It might be well have to ignore some tests. There is a config struct fir the integration tests you could disable tests with, or we could add a new feature flag for the backend NoUpdates or something like that.

**Comment by anjor** [comment]
Another update, that's the only test that's failing now.

![Screenshot 2022-12-19 at 11 32 14 AM](https://user-images.githubusercontent.com/1911631/208417020-7a287ff2-24dd-4090-82dc-d10b5f30d0c2.png)


Is the config where I can disable tests the yaml file -- https://github.com/rclone/rclone/blob/master/fstest/test_all/config.yaml ?

**Comment by anjor** [comment]
@ncw PR is up. I tried rebasing the first one to clean up commit history, but that was proving to be quite difficult; so ended up opening the 2nd PR.

Let me know if I can assist in any way to get this over the finish line. And thank you for your incredible patience and help!

**Comment by anjor** [comment]
Yes, `FsPutFiles` is still failing, but there is an[ open PR](https://github.com/application-research/estuary/pull/727) on estuary to allow for updates. I am hoping to get that merged + released soon, at which point I expect that `FsPutFiles` will succeed.

In case it doesn't, how do you propose using the Feature flag to ignore the test? Like you said, right now I have ignored it in `config.yaml`.


Once we resolve the above point of `FsPutFiles` I can set up an account for you for the integration tester.

**Comment by anjor** [comment]
@ncw i would rather not have it as a feature, and just ignore the test for now in config.yaml. Because I suspect we will support updating files in the future.

Would you be ok with that?

**Comment by ncw** [comment]
> @ncw i would rather not have it as a feature, and just ignore the test for now in config.yaml. Because I suspect we will support updating files in the future.
> 
> Would you be ok with that?

Sounds fine for the moment.

> I just sent you an invite to estuary for the integration tester as well.

Got it, thanks.

**Comment by ncw** [comment]
Is the PR ready for me to test now?

**Comment by anjor** [comment]
Yes it is. Thanks.

**Comment by anjor** [comment]
Sorry about the radio silence here. We are making some big changes on the estuary backend, and I didn't want to conflate the efforts.

The current plan on our end is to ship the new version by late April. Someone on the team will pick this work up soon after.

@lucroy @jlogelin

**Comment by anjor** [comment]
I am going to close this out as estuary is now deprecated.
